### ISSUE TITLE
Enable inline multi-device model tuning via `post_training` API

### ISSUE THREAD

**Issue by JamesKunstle** [meta-llama/llama-stack#1427]
### 🚀 Describe the new functionality needed

The existing model tuning implementation is uses Torchtune and is single-device LoRA for Llama models. This is an excellent representative example for a tuning workflow via the LLS API.
A next reasonable step is to implement inline multi-card full precision tuning for LLS instances deployed on >1x accelerator hosts.

### 💡 Why is this needed? What if we don't build it?

This feature gives the user more flexibility and performance when tuning their models. This would support faster tuning, tuning larger models in higher precision, and higher saturation of host compute resources.

### Other thoughts

This feature would likely require some additional sophistication in the following areas:

1. Job management: A full-precision tuning run can take a long time and use all available resources.
2. Data artifact management: Checkpoints saved during tuning need to be managed logically.
3. Cache management: Saving checkpoints takes disk space- how can we mitigate disk use collisions?

Early implementations can relax these ergonomic challenges but they should be kept in mind for the future.

**Comment by JamesKunstle** [comment]
I'm currently working on this! I created this issue so we could discuss.

**Comment by SLR722** [comment]
That's a great add-on for post training! Happy to help you with this. Keep me posted!

**Comment by JamesKunstle** [comment]
In a first prototype, I wanted to extend the existing Torchtune impl. I made these assumptions:

```
1. Dataset expected to be basic "openai" chat format. No schema translations in first impl.
	1. Each sample is {messages: [{role, content}]}
	2. Use torchtune `chat_dataset` to create `Messages` IR.
3. Using Torchtune utilities to preprocess dataset and mask input sections of sample (system, user) w/ `train_on_input=False`. 
4. Preprocessed dataset is written to tempfile. Each sample will have `input_ids`, `mask`, and `labels`. 
5. Each distrib rank process reads entire dataset; sampler is native Torch distributed sampler.
6. Batches are naively collated (axis-0, no packing, right-padded). 
```

I assumed that I could use the torchtune data preprocessing utilities with an HF tokenizer. This isn't supported in v0.5 Torchtune (looks like there's an HF tokenizer wrapper available on `main`: https://pytorch.org/torchtune/main/basics/tokenizers.html#using-hugging-face-tokenizers).

I'm going to make an alternative plan and implement an HF-utility-based tuning loop for some additional flexibility. I'm updating my assumptions to the following:

1. Input dataset expected to be "openai" chat format ([messages: [{role, content}]]) only. Will implement schema translation later.
2. Model to tune will be specified by _reference_ to HF repo. e.g. `meta-llama/Meta-Llama-3.1-8B-Instruct` or `ibm-granite/granite-3.1-8b-instruct`. Model modules will be loaded via `transformers.AutoModelForCausalLM`.
3. Tokenizer will be loaded via `transformers.AutoTokenizer.from_pretrained(model_ref, use_fast=True)`
4. Preprocessing will halt if `current_tokenizer.apply_chat_template(sample, tokenize=True)` fails. Implicitly, this means that the selected tokenizer must already be set up for an Instruct / Chat -tuned model. In the future, we'll let users specify their own chat template / tokenizer model reference so they can use any base tokenizer, with a specified chat template, to tune any base model.
5. Samples aren't input-loss-masked. Some libraries (torchtune, axolotl) enable this masking by default, while others (hf trainer) don't do this. I haven't found conclusive evidence that non-assistant masking is a _requirement_ so this'll be left to a near future optimization iteration.
6. Batches are naively collated. This will be slower and less efficient than using packed samples, but we can improve that in a near future optimization.
7. Batches are sampled using naive rank-based distributed sampling. No dense batch packing or optimized sampling for packing. Can optimize this in the future.
8. Distributed training uses FSDP2. Traditional throughput configuration will be applied as default: [activation checkpointing, backward_pre, etc.]

**Comment by SLR722** [comment]
Thanks for the detailed plan! I curious why you need a HF-utility-based tuning loop for some additional flexibility instead of extending the current torchtune based training loop to distributed?

**Comment by JamesKunstle** [comment]
That's a reasonable question. I have some fairly human reasons for going this direction:
1. I know how to write an HF[ datasets + tokenizers + transformers ] + Torch[ FSDP ] training loop.
2. The current `torchtune` tokenizers impl doesn't support HF Tokenizers directly (it supports a few models like Qwen, Mistral, Gemini, Llama, but not my current model focus, Granite) so I'd have to write a `torchtune` messages wrapper around arbitrary HF tokenizers that get loaded (If you know how to get around this ergonomically I'd be very interested).
3. I think having an HF example will map to another subset of user experiences and encourage some more hacking in the post_training API space.

Would be very happy to discuss further + consider changing course if you have strong convictions- you're the original implementor so I'm interested in the direction you'd want to take.

**Comment by SLR722** [comment]
That's fair! Do you plan to create a new inline HF provider? If yes, besides tokenizes, what other components (such as model implementation, distributed training, data processing etc) in the training loop you plan to use? torchtune or HF trainer?

**Comment by JamesKunstle** [comment]
Yeah here's a sketch of what I'm thinking:

```python3
## runs once in BackgroundTask or something when endpoint invoked.

def render_template_and_tokenize(sample):
    ...
    return {
        "input_ids" ...,
        "attention_mask": ...,
        "labels": ...,
        "loss_mask": ...,
    }
    
model_config = AutoConfig.from_pretrained(model_name_or_path)

if model_config not in constants.VERIFIED_ARCHITECTURES:
    print(f"{model_config} model architecture support is BEST EFFORT. This architecture hasn't been explicitly verified.")


tok = AutoTokenizer.from_pretrained(model_name_or_path, use_fast=True)
if tok.default_chat_template is None:
    raise RuntimeError(f"Tokenizer for ({model_name_or_path}) does not have a `default_chat_template`. This provider requires chat templates to be prepared before tuning starts.")

dataset = load_dataset("json", dataset_buffer_from_datasetio_api)
validate_dataset(dataset) # raises RuntimeError or something if dataset isn't OpenAI chat-formatted.
dataset_prepared = dataset.map(
    render_template_and_tokenize,
    batched=True
)

data_file = current_tempfile / "data.jsonl"
dataset_prepared.to_json(path=data_file)

# call auxiliary SPMD file w/ torchrun so we get RANK / LOCAL_RANK / WORLD_SIZE in environment for free
# will have to keep this PID in path function cache so user can refer to it later to get training status, etc.
pid = asyncio.create_subprocess_shell(f"torchrun <params> ./distributed_train.py --data-file {data_file} <other args>")


## in `distributed_train.py`

data = load_dataset("json", path=data_file)
init_torch_distributed_utils()

model = AutoModelForCausalLM.from_pretrained(model_name_or_path)

sampler = torch.utils.data.DistributedSampler(dataset=data, num_replicas=os.getenv("WORLD_SIZE"), rank=os.getenv("LOCAL_RANK"))
dataloader = torch.utils.data.DataLoader(
    dataset=data,
    sampler=sampler,
    shuffle=True,
    ...
)

# run training 
```

There are quite a few ergonomic challenges w/ this, including:
1. Whether models should come from a `models` and `modelsio` resource (that'd be my preference in the longer term) so that we could have the backend be torchrun, s3, oci, hf, etc.
2. How users should get checkpoints back after tuning,
3. how GPUs should be "locked" explicitly during long-running training operation
4. ... many others.

But I think this is a pretty boilerplate first pass.

**Comment by SLR722** [comment]
Thanks for the detailed plan! Overall, it makes sense. Since you are using HF trainer and plan to call it HF provider, for the distributed training part, do you wanna use Accelerate + FSDP?

**Comment by JamesKunstle** [comment]
I won't be using HF Trainer directly- it'll be FSDP directly, though. I'd choose not to use Accelerate because I'd like to use Torch-native stuff (FSDP) and Accelerate is a set of additional deps we could avoid (namely, DeepSpeed and its family).

**Comment by SLR722** [comment]
By reading their doc, seems Accelerate support native FSDP? https://huggingface.co/docs/accelerate/usage_guides/fsdp

My concern here is if we combined pieces from difference places, if HF is the appropriated provider name.

**Comment by JamesKunstle** [comment]
That's a fair point. I'd consider this "HF and Torch" thematically- Accelerate gives us _both_ FSDP and DeepSpeed, and I'd prefer to build this w/o Accelerate as the wrapper around the distributed training impl- I don't intend to expose the DeepSpeed bits so Accelerate feels like overkill. At least for a first iteration. 

i.e. the utils that makes this path easy to understand and conventional are [datasets, tokenizers, transformers] from HF, and the training is all Torch-native. If we add [accelerate] it's more HF stuff, but I don't think we need to grab _all_ the HF stuff that's available 😁 Very happy to try it a different way if you feel strongly, though.

We maintain an Accelerate loop [in our repo here](https://github.com/instructlab/training/blob/main/src/instructlab/training/main_ds.py) and it's very nice, but it's unnecessary if we're just doing FSDP directly. Accelerate doesn't really save us steps, it's just nice when wrapping DeepSpeed _and_ FSDP as we do, serving both backends.
